### PR TITLE
bpo-1812: Fix newline conversion when doctest.testfile loads from a package whose loader has a get_data method.

### PR DESCRIPTION
**Pull request by pdonis**
[bpo-1812](https://bugs.python.org/issue1812): Fix newline conversion when doctest.testfile loads from a package whose loader has a get_data method.

This pull request fixes the newline conversion bug originally reported in issue 1812. When that issue was originally submitted, the open builtin did not default to universal newline mode; now it does, which makes the issue fix simpler, since the only code path that needs to be changed is the one in doctest._load_testfile where the file is loaded from a package whose loader has a get_data method. Commit b6ad8ee contains the change that implements the fix.

This pull request also adds a test that verifies that doctest.testfile does the newline conversion correctly when loading from a package whose loader has a get_data method. This test fails (a ValueError is raised for inconsistent whitespace) unless the change in commit b6ad8ee is applied. Commit dd02845 contains the changes that add the test; the test is added to the test_lineendings docstring along with supporting comments.

Also, the test_lineendings docstring is missing a test to check for correct handling of the old-style Mac newline \r. This pull request adds that case to the original test_lineendings docstring. Commit da35562 contains the changes that add this test and edit the surrounding comments accordingly.

Finally, this pull request contains commit 22094c8 that adds my name to the MISC/Acks file.

https://bugs.python.org/issue1812

### CODE REVIEW

File: Lib/test/test_doctest.py
Comment (zware) [inline Lib/test/test_doctest.py]: It would be nice to break up the long line here; it should be doable to split it up in some logical places and allow implicit concatenation to reconstruct it:

```suggestion
    ...     f.write(
    ...         b'Test:\r\n\r\n'
    ...         b'  >>> x = 1 + 1\r\n\r\n'
    ...         b'Done.\r\n'
    ...         b'Test:\n\n'
    ...         b'  >>> x = 1 + 1\n\n'
    ...         b'Done.\n'
    ...         b'Test:\r\r'
    ...         b'  >>> x = 1+ 1\r\r'
    ...         b'Done.\r'
    ...     )
```

File: Lib/test/test_doctest.py
Comment (zware) [inline Lib/test/test_doctest.py:2697]: It would be better to do this in a context manager just to be very sure we don't miss the cleanup.  There are some helpers in `test.test_importlib.util` that may be helpful here, though I confess I don't understand the inner workings of `importlib` well enough to know exactly what we need here :).  On the other hand, perhaps we can just pass in a mock object for `package` and avoid messing with global import state at all.

**Review by zware — COMMENTED** [review]
Overall, this is an incredibly clean patch, and you have the patience of a saint to have shepherded this along for 12 years!  I have a style nit and a cleanup suggestion, but this is probably acceptable as is.

Thanks for the patch, and hopefully we'll finally get this landed!

**Comment by aeros** [comment]
This seems like it should be backported to 3.6+, no? It's fixing an internal function and making some changes to the tests, so I don't think there's any backwards compatibility concerns to worry about. I'll add the backport labels for now, but feel free to remove them if needed.

**Comment by zware** [comment]
3.6 is in security-fix-only mode, so this definitely will not be backported that far.

**Comment by aeros** [comment]
Oh oops, for some reason I had forgotten that 3.6 was a security-only branch. 3.7+ would be better in that case.

File: Lib/test/test_doctest.py
Comment (pdonis) [reply Lib/test/test_doctest.py:2697]: I looked through `test.test_importlib.util` and didn't see anything that I thought I could use. I'm also not confident enough in my knowledge of how packages work to try to construct a mock package object that won't cause some other problem.

The context manager suggestion is a good one. It looks straightforward enough, I will test that and push an update if I can get it working.

Comment (pdonis) [reply Lib/test/test_doctest.py:2697]: The context manager turned out to be very simple; `contextlib.contextmanager` is my friend. :-)

There are a total of three new commits:

Commit efabd42 does the breakup of the long line (I just committed your suggested change, @zware).

Commit 1d56c05 fixes a typo in the breakup of the long line.

Commit e1f0a2f implements the context manager.

Comment (pdonis) [reply Lib/test/test_doctest.py:2697]: @zware Thanks for the kudos, btw!

**Comment by pdonis** [comment]
Any update on this? Should I be looking at the 3.7 and 3.8 branches for backporting?

**Comment by aeros** [comment]
> Any update on this?

I believe at this point we're just waiting for a follow-up review from @zware.

> Should I be looking at the 3.7 and 3.8 branches for backporting?

The labels I added will trigger @miss-islington to attempt to open backport PRs to the 3.8 and 3.7 branches after this one is merged. It typically works without conflicts, but if there are any, you may have to resolve them manually with `git cherry-pick` (or the tool on PyPI, `cherry-picker`).

File: Lib/doctest.py
Comment (ammaraskar) [inline Lib/doctest.py:231]: Instead of trying to replicate the behavior of universal new lines, perhaps we can combine this and the `decode` routine above with something like:

```python
>>> file_contents = b"line1\r\nline2\nline3\rline4"
>>> io.TextIOWrapper(io.BytesIO(file_contents), encoding='utf8').read()
'line1\nline2\nline3\nline4'
```

File: Lib/test/test_doctest.py
Comment (ammaraskar) [inline Lib/test/test_doctest.py:2700]: I think some of this and the code above can be folded to use stuff from https://github.com/python/cpython/blob/master/Lib/test/test_importlib/fixtures.py like:

* https://github.com/python/cpython/blob/fd5116c0e77aec05f67fb24f10562ac920648035/Lib/test/test_importlib/fixtures.py#L71-L83

* https://github.com/python/cpython/blob/fd5116c0e77aec05f67fb24f10562ac920648035/Lib/test/test_importlib/fixtures.py#L50-L56

Comment (pdonis) [reply Lib/test/test_doctest.py:2700]: The OnSysPath fixture appears to be designed for a unittest test case, not doctest. A standalone version of the add_sys_path context manager might work if combined with install_finder; however, I notice that neither of those clears the path importer cache, and neither one removes the imported test module from sys.modules when the test is done. I don't know enough about the internals of the test framework to know if leaving out those last two items is OK; I put them in to make sure that everything that could possibly be mutated by this particular test was restored afterwards.

File: Lib/doctest.py
Comment (pdonis) [reply Lib/doctest.py:231]: I don't know enough about the internals of TextIOWrapper and BytesIO to know if this would risk breaking anything else. My intent was to make the minimal possible fix to the existing code, and since the existing code used the string replace operation, I just fixed that up to do the correct replacements.

I also don't know if the read operation on TextIOWrapper, or the BytesIO constructor, might throw different exceptions on malformed data than the simple decode and string replace operations would.

**Comment by pdonis** [comment]
@aeros Thanks for the update and for explaining how the backport process works.

**Comment by pdonis** [comment]
@zware is there any update on this?

File: Lib/test/test_doctest.py
Comment (brettcannon) [inline Lib/test/test_doctest.py]: This is misnamed as this is implementing a finder and a loader, so call it an importer.

File: Lib/test/test_doctest.py
Comment (brettcannon) [reply Lib/test/test_doctest.py:2697]: You want `test.test_importlib.util.import_state()`: https://github.com/python/cpython/blob/master/Lib/test/test_importlib/util.py#L183

File: Lib/test/test_doctest.py
Comment (pdonis) [reply Lib/test/test_doctest.py]: I'll push an update changing the name to TestImporter.

Comment (pdonis) [reply Lib/test/test_doctest.py]: Commit bb836d7 does the name change.

**Comment by pdonis** [comment]
Any update on merging this?

**Comment by brettcannon** [comment]
I refreshed the review request for @zware to see if he has time to review again.

File: Lib/doctest.py
Comment (zware) [reply Lib/doctest.py:231]: The concern about possibly changing exceptions is enough for me to want to stick with the simple fix here.

File: Lib/test/test_doctest.py
Comment (zware) [reply Lib/test/test_doctest.py:2697]: @pdonis, did you see @brettcannon's message here?  I think there is some opportunity to clean up the clean up code a bit, but on the other hand that can be done in a follow-on PR.

**Review by zware — APPROVED** [review]
I wouldn't mind seeing the cleanup code cleaned up a bit in a separate PR, but as nobody has pointed out any actual *problems* with it as is, let's go ahead.

**Comment by pdonis** [comment]
@zware @brettcannon Thanks so much for getting this merged!

**Pull request merged**
